providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java

```diff
@@ -102,43 +105,54 @@ public void shutdown() throws Exception {
         }
     }
 
-    /**
-     * Resets internal state for retrying establishing the stream.
-     */
-    public void resetRetryConnection() {
-        this.eventStreamAttempt = 1;
-        this.eventStreamRetryBackoff = this.startEventStreamRetryBackoff;
-    }
-
     /**
      * Provide the object that can be used to resolve Feature Flag values.
+     *
      * @return a {@link ServiceGrpc.ServiceBlockingStub} for running FF resolution.
      */
     public ServiceGrpc.ServiceBlockingStub getResolver() {
         return serviceBlockingStub.withDeadlineAfter(this.deadline, TimeUnit.MILLISECONDS);
     }
 
-    private void startEventStream() {
-        StreamObserver<Schema.EventStreamResponse> responseObserver =
-                new EventStreamObserver(this.cache, this.stateConsumer, this::restartEventStream);
-        this.serviceStub
-                .eventStream(Schema.EventStreamRequest.getDefaultInstance(), responseObserver);
-    }
+    private void observeEventStream() {
+        while (this.eventStreamAttempt <= this.maxEventStreamRetries) {
+            StreamObserver<Schema.EventStreamResponse> responseObserver =
+                    new EventStreamObserver(sync, this.cache, this::grpcStateConsumer);
+            this.serviceStub.eventStream(Schema.EventStreamRequest.getDefaultInstance(), responseObserver);
+
+            try {
+                synchronized (sync) {
+                    sync.wait();
+                }
+            } catch (InterruptedException e) {
+                log.warn("interruption while waiting for condition", e);
+                return;
+            }
+
+            this.eventStreamAttempt++;
+            this.eventStreamRetryBackoff = 2 * this.eventStreamRetryBackoff;
 
-    private void restartEventStream() {
-        this.eventStreamAttempt++;
-        if (this.eventStreamAttempt > this.maxEventStreamRetries) {
-            log.error("failed to connect to event stream, exhausted retries");
-            this.stateConsumer.accept(ProviderState.ERROR);
-            return;
+            try {
+                Thread.sleep(this.eventStreamRetryBackoff);
```

Comment:
Can consider adding [Jitter](https://aws.amazon.com/builders-library/timeouts-retries-and-backoff-with-jitter/).

Reply:
> If this can be called from multiple threads, and the other thread which wait on this object is during wait(), then this here may be stuck on synchronized ?
Is it possible ?

As explained here [^1], we do not use multiple threads. We only have 1-1 relationship from caller to async stream handler.

Given sync block is the standard way to deal with `wait` & `notify` [^2], we shall not face with deadlocks. 

Regarding adding jitter, I think this can be considered a follow-up 

[^1]: https://github.com/open-feature/java-sdk-contrib/pull/398#discussion_r1296076334
[^2]: https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#notify--

Reply:
Given that adding jitter is a simple change, I added this to our logic. See commit - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/d28bdcd679fef903566697677b7d5a8797417c18